server/src/main/java/org/cloudfoundry/identity/uaa/oauth/UaaTokenStore.java

```diff
@@ -245,31 +243,35 @@ public TokenCode mapRow(ResultSet rs, int rowNum) throws SQLException {
             String userid = rs.getString(pos++);
             String client_id = rs.getString(pos++);
             long expiresat = rs.getLong(pos++);
-            Timestamp created = rs.getTimestamp(pos++);
+            Instant created = rs.getTimestamp(pos++).toInstant();
             byte[] authentication = rs.getBytes(pos++);
-            return createTokenCode(code, userid, client_id, expiresat, created, authentication);
+
+            if (expiresat == 0) {
+                return new LegacyTokenCode(code, userid, created, client_id, authentication);
+            } else {
+                return new NewTokenCode(code, userid, Instant.ofEpochMilli(expiresat), client_id, authentication);
+            }
         }
     }
 
-    public TokenCode createTokenCode(String code, String userId, String clientId, long expiresAt, Timestamp created, byte[] authentication) {
-        return new TokenCode(code,userId,clientId,expiresAt,created,authentication);
+    public TokenCode createTokenCodeForTesting(String code, String userId, String clientId, Optional<Instant> expiresAt, Instant created, byte[] authentication) {
```

Comment:
do want still go with Optional in a public method ?

Reply:
I don't love it, but as I said in my other message, it's a public method only for testing purposes. Actually I may just change it to package private, that should help a bit.